src/Microsoft.Extensions.Correlation/Internal/HttpDiagnosticListenerObserver.cs

```diff
@@ -13,69 +13,79 @@ namespace Microsoft.Extensions.Correlation.Internal
 {
     internal class HttpDiagnosticListenerObserver : IObserver<KeyValuePair<string, object>>
     {
-        private readonly EndpointFilter filter;
-        private readonly CorrelationConfigurationOptions.HeaderOptions headerMap;
         private readonly DiagnosticListener listener;
-        public HttpDiagnosticListenerObserver(EndpointFilter filter, CorrelationConfigurationOptions.HeaderOptions headerMap,
-            DiagnosticListener listener)
+        private PropertyFetch requestFetcher;
+        private PropertyFetch responseFetcher;
+        private PropertyFetch requestTimestampFetcher;
+        private PropertyFetch responseTimestampFetcher;
+
+        public HttpDiagnosticListenerObserver(DiagnosticListener listener)
         {
-            this.filter = filter;
-            this.headerMap = headerMap;
             this.listener = listener;
+            requestFetcher = null;
+            requestTimestampFetcher = null;
+            responseFetcher = null;
+            responseTimestampFetcher = null;
         }
 
         public void OnNext(KeyValuePair<string, object> value)
         {
             if (value.Value == null)
                 return;
 
-            //Request and Response events handling id done for 2 reasons
-            //1. inject headers
-            //2. We need to notify user about outgoing request;  user may want to log outgoing requests because 
-            //  - downstream service logs may not be available (it's external dependency or not instumented service)
-            //  - user may want to create visualization for operation flow and need parent-child relationship between requests:
-            //       e.g. service calls other service multiple times within the same operation (because of retries or business logic)
-            //       so activity id should be logged on this service and downstream service to uniquely map one request to another, having the same correlation id
-            //  - user is interested in client time difference with server time
-
             if (value.Key == "System.Net.Http.Request")
             {
-                var request = (HttpRequestMessage)value.Value.GetProperty("Request");
+                if (requestFetcher == null)
+                    requestFetcher = PropertyFetch.FetcherForProperty(value.Value.GetPropertyInfo("Request"));
```

Comment:
Before we check in for real, I want to change System.Net.Http so that it exposes the name of the type that contains the payload for these Write() calls.   That way you can simply cast the value to that type and fetch it (which is the most efficient).

So we really should not need the PropertyFetch code in here.   however I would not mind if we added this code to DiagnosticSource (in this module we were willing to take a dependency on System.Net.Http, but sometimes we are not willing to do this, and PropertyFetcher is useful in that case).    

As an aside, I would probably make the PropertyFetch constructor take the property Name, and then do the lazy Init at fetch time (when you have 'this' pointer (and thus the type)).   It is just as efficient and nicer to use.